python testing: Per node endpoint fixes

cecille opened the pull request:
- change name of decorator to hopefully be more obvious what it does
- change to force the use of a specific endpoint so that this is easier to use in the TH

Test: Running against CI, also see TestDecorators for new tests of override

cecille committed:
Change name of decorator

cecille committed:
Add a force option for per endpoint tests

tcarmelveilleux commented on src/python_testing/matter_testing_support.py:
```diff
@@ -1859,15 +1859,21 @@ def has_feature(cluster: ClusterObjects.ClusterObjectDescriptor, feature: IntFla
 
         Returns a list of endpoints on which the test should be run given the accept_function for the test.
     """
     wildcard = await self.default_controller.Read(self.dut_node_id, [()])
```
Would reading only the descriptor on each endpoint suffice? Would be a much faster read.

cecille replied:
yeah, probably would be faster. Thanks. Will change that.

cecille replied:
oh, no, wait, it needs to be able to determine the feature map and attribute list for each cluster as well. So I need the full wildcard.

cecille replied:
Maybe descriptor and globals would do though

cecille replied:
I'm going to land this change separately as a follow up PR since it requires a change to the Read function that I'd like to ensure gets proper feedback.

tcarmelveilleux commented on src/python_testing/matter_testing_support.py:
```diff
@@ -1859,15 +1859,21 @@ def has_feature(cluster: ClusterObjects.ClusterObjectDescriptor, feature: IntFla
     wildcard = await self.default_controller.Read(self.dut_node_id, [()])
-    return [e for e in wildcard.attributes.keys() if accept_function(wildcard, e)]
+    matching = [e for e in wildcard.attributes.keys() if accept_function(wildcard, e)]
+    forced_endpoint = self.user_params.get('force_endpoint', None)
```
Suggest logging when a forced endpoint is used.

cecille replied:
added.

tcarmelveilleux commented on src/python_testing/TC_LVL_2_3.py:
```diff
@@ -81,7 +81,7 @@ def steps_TC_LVL_2_3(self) -> list[TestStep]:
                 ]
 
-    @per_endpoint_test(has_cluster(Clusters.LevelControl))
+    @run_for_each_matching_endpoint(has_cluster(Clusters.LevelControl))
```
This should have docs about how `--int-arg forced_endpoint:N` is how you run on only one endpoint at a time for testing.

cecille replied:
Added to the assertion message.

cecille committed:
fix test

cecille committed:
Add warning on force endpoint

cecille committed:
Disallow --endpoint with automatic endpoint selection

restyled-commits committed:
Restyled by isort

cecille committed:
merge conflict fix

cecille committed:
Don't automatically override endpoint in MockRunner

restyled-commits committed:
Restyled by isort

cecille committed:
Fix endpoint default on command line

cecille committed:
Merge branch 'master' into per_node_endpoint_fixes

cecille committed:
Merge remote-tracking branch 'upstream/master' into per_node_endpoint_fixes

cecille committed:
Going the other way on this

With the way the ATLs expect the test files currently, this is
causing issues with submission. The ATLs aren't familiar with
how to read these files and it's making cert challenging.

Instead, we're going to gate or skip tests PER ENDPOINT such that
the --endpoint flag can be set. This also (in theory) lets the tests
run alongside other tests, where the --endpoint is specified.

The auto-runner was too much, too soon. It's great for CI, but
we don't have TH features to select these tests, nor ways to
allow the file submission. Let's fall back to once per endpoint.
It's still better, we're not relying on PICS.

cecille committed:
Update comments.

cecille committed:
Update all tests

cecille committed:
more things

restyled-commits committed:
Restyled by isort

cecille committed:
fix singleton test

cecille committed:
fix var

bzbarsky-apple commented on src/python_testing/matter_testing_support.py:
```diff
@@ -1896,31 +1881,69 @@ def has_feature(cluster: ClusterObjects.ClusterObjectDescriptor, feature: IntFla
-def per_endpoint_test(accept_function: EndpointCheckFunction):
-    """ Test decorator for a test that needs to be run once per endpoint that meets the accept_function criteria.
+def run_if_endpoint_matches(accept_function: EndpointCheckFunction):
+    """ Test decorator for a test that needs to be run only if the endpoint that meets the accept_function criteria.
```
```suggestion
    """ Test decorator for a test that needs to be run only if the endpoint meets the accept_function criteria.
```

bzbarsky-apple reviewed (approved):
I did not review TestDecorators.py carefully, and might have missed subtleties in the core guts in matter_testing_support.py...

cecille committed:
Update src/python_testing/matter_testing_support.py

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

fabiowmm reviewed (approved):
LGTM.

cecille committed:
Merge remote-tracking branch 'upstream/master' into per_node_endpoint_fixes

cecille committed:
Merge remote-tracking branch 'origin/per_node_endpoint_fixes' into per_node_endpoint_fixes

restyled-commits committed:
Restyled by isort

cecille committed:
add endpoint flags
